src/Notes/WooCommercePayments.php

```diff
@@ -128,22 +135,58 @@ protected static function is_installed() {
 	}
 
 	/**
-	 * Install WooCommerce Payments when note is actioned.
+	 * Install and activate WooCommerce Payments.
 	 *
-	 * @param Note $note Note being acted upon.
+	 * @return boolean Whether the plugin was successfully activated.
 	 */
-	public function install( $note ) {
-		if ( self::NOTE_NAME === $note->get_name() && current_user_can( 'install_plugins' ) ) {
-			$install_request = array( 'plugins' => self::PLUGIN_SLUG );
-			$installer       = new \Automattic\WooCommerce\Admin\API\Plugins();
-			$result          = $installer->install_plugins( $install_request );
+	private function install_and_activate_wcpay() {
+		$install_request = array( 'plugins' => self::PLUGIN_SLUG );
+		$installer       = new \Automattic\WooCommerce\Admin\API\Plugins();
+		$result          = $installer->install_plugins( $install_request );
+		if ( is_wp_error( $result ) ) {
+			return false;
+		}
 
-			if ( is_wp_error( $result ) ) {
-				return;
-			}
+		$activate_request = array( 'plugins' => self::PLUGIN_SLUG );
+		$result           = $installer->activate_plugins( $activate_request );
+		if ( is_wp_error( $result ) ) {
+			return false;
+		}
+
+		return true;
+	}
 
-			$activate_request = array( 'plugins' => self::PLUGIN_SLUG );
-			$installer->activate_plugins( $activate_request );
+	/**
+	 * Install & activate WooCommerce Payments plugin, and redirect to setup.
+	 */
+	public function install_on_action() {
+		// TODO: Need to validate this request more strictly since we're taking install actions directly?
+		/* phpcs:disable WordPress.Security.NonceVerification */
+		if (
```

Comment:
ToS acceptance is tricky part here. At the moment we kind of can't ensure the action was taken by merchant clicking and accepting the ToS. Maybe before redirecting we need to check ToS acceptance somehow? Related to Automattic/woocommerce-payments#887

Reply:
I'm not caught up on every nuance there, but as far as I can tell, we may be able to address that issue by noting the acceptance data immediately before redirect (and continue to assume that ToS acceptance was presented in one way or another before then)? cc @marcinbot @allendav @thenbrent

In any case, I don't think this redirect would behave any different in that regard from the WCPay task ones, one of which recently had a ToS acceptance added in https://github.com/woocommerce/woocommerce-admin/pull/5085, prompted by that thread – paJDYF-OD-p2#comment-3315.